cmd/snap/cmd_run.go

```diff
@@ -1156,55 +1164,88 @@ var osGetpid = os.Getpid
-			logger.Debugf("snap refreshes will not be postponed by this process")
-			return nil
-		}
-	}
-	if err != nil {
-		return err
-	}
+
 	// We ask the kernel for a random UUID. We need one because each transient
 	// scope needs a unique name. The unique name is comprosed of said UUID and
 	// the snap security tag.
 	uuid, err := randomUUID()
 	if err != nil {
 		return err
 	}
+
 	// Enforcing uniqueness is preferred to reusing an existing scope for
 	// simplicity since doing otherwise and joining an existing scope has
 	// limitations:
 	// - the originally started scope must be marked as a delegate, with all consequences.
 	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	//
 	// TODO: invert the arguments so that security tag comes first, UUID
 	// second. This will result in nicer scope names but needs to be
 	// coordinated with code in sandbox/cgroup and with the spread tests
 	// exercising that.
 	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
 	pid := osGetpid()
+
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	isSessionBus := true
+	conn, err := dbusSessionBus()
+	if err != nil {
+		logger.Debugf("Session bus is not available: %s", err)
+		if osGetuid() == 0 {
+			logger.Debugf("Falling back to system bus")
+			isSessionBus = false
+			conn, err = dbusSystemBus()
+			if err != nil {
+				logger.Debugf("System bus is not available: %s", err)
+			}
+		}
+	}
+
+	// Session or system bus might be unavailable. To avoid being fragile
+	// ignore all errors when establishing session bus connection to avoid
+	// breaking user interactions. This is consistent with similar failure
+	// modes below, where other parts of the stack fail.
+	//
+	// Ideally we would check for a distinct error type but this is just an
+	// errors.New() in go-dbus code.
+	if err != nil {
+		return errCannotTrackProcess
+	}
+
+tryAgain:
 	// Create a transient scope by talking to systemd over DBus.
 	if err := doCreateTransientScope(conn, unitName, pid); err != nil {
+		switch err {
+		case errDBusUnknownMethod:
+			return errCannotTrackProcess
+		case errDBusSpawnChildExited:
+			// We cannot activate systemd --user for root, try the system
+			// bus as a fallback.
+			if isSessionBus && osGetuid() == 0 {
+				logger.Debugf("Cannot activate systemd --user on session bus, falling back to system bus")
+				isSessionBus = false
+				conn, err = dbusSystemBus()
+				if err != nil {
+					return errCannotTrackProcess
+				}
+				goto tryAgain
```

Comment:
why not an explicit `for { }` loop?

Reply:
I kind of couldn't make it nice. I will still tweak it, now that the test is written, and I can expect it to catch regressions.